### PR TITLE
Add support for switching on float, double, and decimal.

### LINKED ISSUE #12431: Implement switch for float, double, and decimal
The typeswitch (C# 7 pattern-matching) implementation does not currently support switching on values of type float, decimal, and double.

### PR DESCRIPTION
**Pull request by gafter**
The code is a sequence of if-then-else tests using value.Equals(input).
Someday we might do better using divide-and-conquer.

Fixes #12431.

@cston @agocke @jaredpar @dotnet/roslyn-compiler May I please have a couple of reviews for this preview 4 change?

### CODE REVIEW

**Comment by gafter** [comment]
@cston @agocke @jaredpar @dotnet/roslyn-compiler This is the last PR needed to complete the typeswitch implementation for preview 4. May I please have a couple of reviews so I can place it in ask mode?

**Comment by agocke** [comment]
Reviewing now.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 1]
```diff
@@ -391,24 +360,101 @@ private void LowerBasicSwitch(DecisionTree.ByValue byValue)
                 }
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Assert that the special type is not `boolean`? (assuming that all boolean lowering should go through the other method).

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 2]
```diff
@@ -391,24 +360,101 @@ private void LowerBasicSwitch(DecisionTree.ByValue byValue)
...
+                            _loweredDecisionTree.Add(switchStatement);
+                            // The bound switch statement implicitly defines the label noValueMatches at the end, so we do not add it explicitly.
+                            break;
+                        }
+
+                    default:
+                        {
+                            if (byValue.Type.TypeKind == TypeKind.Enum)
+                            {
+                                goto case SpecialType.System_Int32;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Consider breaking the `default` and other code blocks into their own method, then calling the `integer` helper method here and in the `System_String` case.

**Comment by agocke** [comment]
LGTM aside from comments

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: This code does not actually assume or depend on the fact that it is *not* boolean, so such an assertion would be out of place.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: I'll place the assertion into the switch, below.

---
In reply to: [70700915](https://github.com/dotnet/roslyn/pull/12478#discussion_r70700915) [](ancestors = 70700915)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: I'll break out the float/double/decimal code handling. The primitive handling code is only three lines.

---
In reply to: [70701623](https://github.com/dotnet/roslyn/pull/12478#discussion_r70701623) [](ancestors = 70701623)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 3]
```diff
@@ -392,26 +361,120 @@ private void LowerBasicSwitch(DecisionTree.ByValue byValue)
...
+                var builder = ArrayBuilder<BoundStatement>.GetInstance();
+                foreach (var section in rewrittenSections)
+                {
+                    foreach (var boundSwitchLabel in section.SwitchLabels)
+                    {
+                        if (nextLabel != null)
+                        {
+                            builder.Add(_factory.Label(nextLabel));
+                        }
+                        nextLabel = _factory.GenerateLabel("failcase+" + section.SwitchLabels[0].ConstantValueOpt.Value);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: If this label prefix is supposed to be the same for all switch labels in this section, consider calculating the prefix outside the loop.

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs [hunk 4]
```diff
@@ -851,6 +856,11 @@ public BoundLiteral Literal(string value)
             return StringLiteral(stringConst);
         }
 
+        public BoundLiteral Literal(ConstantValue value, TypeSymbol type)
+        {
+            return new CSharp.BoundLiteral(Syntax, value, type) { WasCompilerGenerated = true };
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: Is `CSharp.` qualifier needed?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: As it turns out, there will only be one label in each section for the switch statements that ever get to this translation code. So it won't make any difference.

---
In reply to: [70733585](https://github.com/dotnet/roslyn/pull/12478#discussion_r70733585) [](ancestors = 70733585)

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: No, and neither is this helper needed any longer.

---
In reply to: [70734062](https://github.com/dotnet/roslyn/pull/12478#discussion_r70734062) [](ancestors = 70734062)

**Comment by gafter** [comment]
@MattGertz This is the last piece needed to have a complete implementation of the typeswitch feature. This is proposed to be added to update 4 (now in ask mode), please.

/cc @jaredpar

**Pull request merged**